File: web/js/services/workspaces.js
//     ▄▄ ▄▄ ▄▄  ▄▄▄▄  ▄▄▄▄ ▄▄    ▄▄▄▄▄ ▄▄▄▄
//     ██ ██ ██ ██ ▄▄ ██ ▄▄ ██    ██▄▄  ██▄█▄   Copyright (c) 2026 Julian Storer
//   ▄▄█▀ ▀███▀ ▀███▀ ▀███▀ ██▄▄▄ ██▄▄▄ ██ ██   AGPL-3.0-or-later - see LICENSE

/**
 * Writing to the session's workspace table.
 *
 * The table itself is read from {@link Session#workspaces}, which the load fills
 * and every broadcast replaces. This module is the other direction: the calls
 * that put a row there, move it along, and take it away again. They are here
 * rather than on `Session` because the caller is the provisioning host, not the
 * model — a row exists before anything in the browser is bound to it, and often
 * before the thing it describes exists on disk.
 *
 * Nothing here updates {@link Session#workspaces} itself. Every one of these
 * endpoints broadcasts the whole table, and the session's `workspaces-changed`
 * handler applies it — so a client that also wrote the row locally would be
 * holding a second, briefly-different copy of state the server owns. The
 * returned row is the server's answer to this call, for a caller that needs the
 * assigned id immediately; it is not the table.
 * @module services/workspaces
 */

import { fetchJson } from './http.js';

/**
 * Whether a workspace can be worked in.
 *
 * The one place this is decided on the client. It was written out separately at
 * three surfaces — the binding resolver, the composer chip and the new
 * conversation's picker — and they drifted: the picker checked the state and
 * not the root, so a workspace whose tree had been deleted went on being
 * offered as somewhere to start work, described by the error from the request
 * that failed against it. A predicate spelled out at each call site is a
 * predicate that will differ at one of them.
 *
 * It answers the same question as the server's `WorkspaceLookup.Usable`, which
 * stays the authority — this cannot be shared with Go, so the two are kept
 * deliberately alike and the server refuses again for anything that gets past
 * here. What this is for is not offering the user something that will be
 * refused.
 *
 * Both non-ready states refuse every operation, and a root that is not there
 * refuses them just as completely while saying nothing about itself. Absent
 * availability is treated as available: a row from a server that has not
 * answered yet is not evidence of a missing tree.
 * @param {import('../model/session.js').Workspace|null|undefined} workspace - The row, or nothing.
 * @returns {workspace is import('../model/session.js').Workspace} True when
 *   operations against it would be honoured — and, for a caller about to read
 *   the row, that there is a row to read.
 */
export function isWorkspaceUsable(workspace) {
  if (!workspace) return false;
  return workspace.state === 'ready' && workspace.available !== false;
}

/**
 * A change to one workspace. Every field is optional and an omitted one leaves
 * what is there alone, so a provider can record that it has built one thing
 * without restating a row it has not finished building.
 *
 * `meta` merges key by key rather than replacing, and a `null` value deletes its
 * key. That is what lets two checkpoints — or two windows — write different keys
 * without either erasing the other's.
 * @typedef {object} WorkspacePatch
 * @property {string} [label] - What the UI calls it.
 * @property {string} [root] - Where it is, once that is known.
 * @property {string} [state] - 'provisioning' | 'ready' | 'closed'.
 * @property {Record<string, any>} [meta] - Keys to merge into the provider's own record.
 */

/**
 * Every registered workspace.
 *
 * A viewer has the table already, so this is for the two callers that want what
 * the server holds right now rather than what this window was last told: the
 * reconcile pass, and the setup panel as it opens.
 *
 * It is also the only thing that re-checks whether each root is still there.
 * That makes the call worth making for its effect alone — the answer arrives
 * separately, as the `workspaces-changed` broadcast every viewer gets.
 * @returns {Promise<import('../model/session.js').Workspace[]>} The table.
 */
export async function listWorkspaces() {
  const answer = await fetchJson('/api/session/workspaces');
  return answer?.workspaces ?? [];
}

/**
 * Put a workspace on the table.
 *
 * Registering comes FIRST, before anything is built: a provision that dies half
 * way through then leaves a row saying what was started, rather than an
 * unrecorded half-made tree for its user to find months later. So the usual call
 * names a state of `provisioning` and no root at all.
 *
 * The server names the row unless the caller does. Naming one is for putting
 * back a row a conversation is still bound to — a table lost with its
 * `session.json`, where the binding survives in the conversation's own document
 * and the id is all of it that does. Any other id is refused as already
 * registered.
 * @param {Partial<import('../model/session.js').Workspace>} workspace - The row to create; without an id the server assigns one.
 * @returns {Promise<import('../model/session.js').Workspace>} The row as stored, id included.
 */
export async function registerWorkspace(workspace) {
  const answer = await fetchJson('/api/session/workspaces', { method: 'POST', body: workspace });
  return answer.workspace;
}

/**
 * Change part of one workspace.
 * @param {string} id - The workspace to change.
 * @param {WorkspacePatch} patch - What to change about it.
 * @returns {Promise<import('../model/session.js').Workspace>} The row as it now stands.
 */
export async function patchWorkspace(id, patch) {
  const answer = await fetchJson(`/api/session/workspaces/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body: patch
  });
  return answer.workspace;
}

/**
 * Tombstone a workspace: the row stays and the id keeps resolving, so a
 * conversation still bound to it is told it was closed rather than told it is
 * unknown. That second message belongs to a binding that is genuinely stale.
 * @param {string} id - The workspace to close.
 * @returns {Promise<import('../model/session.js').Workspace>} The closed row.
 */
export async function closeWorkspace(id) {
  const answer = await fetchJson(`/api/session/workspaces/${encodeURIComponent(id)}/close`, {
    method: 'POST'
  });
  return answer.workspace;
}

/**
 * Remove a row outright.
 *
 * What rolling back a provision does, and the only case where forgetting is the
 * right answer: the workspace was never finished, so nothing can have been bound
 * to it. Anything a user has worked in is {@link closeWorkspace}'s to deal with.
 * @param {string} id - The workspace to forget.
 * @returns {Promise<void>} When it is gone.
 */
export async function unregisterWorkspace(id) {
  await fetchJson(`/api/session/workspaces/${encodeURIComponent(id)}`, { method: 'DELETE' });
}

/**
 * Ask to be the client that reconciles the table against what is on disk.
 *
 * Answers yes at most once per run of the server, and the claim is spent by
 * asking. Reconciling means asking each provider what it can find, and providers
 * are extensions living in the browser — but there is no leader among clients,
 * so without this every open window would run it, and two of them would race
 * each other's destructive git commands over the same trees.
 * @returns {Promise<boolean>} Whether this client should reconcile.
 */
export async function claimWorkspaceReconcile() {
  const answer = await fetchJson('/api/session/workspaces/reconcile', { method: 'POST' });
  return answer?.reconcile === true;
}

File: cmd/juggler/server/handlers/workspaces.go
//     ▄▄ ▄▄ ▄▄  ▄▄▄▄  ▄▄▄▄ ▄▄    ▄▄▄▄▄ ▄▄▄▄
//     ██ ██ ██ ██ ▄▄ ██ ▄▄ ██    ██▄▄  ██▄█▄   Copyright (c) 2026 Julian Storer
//   ▄▄█▀ ▀███▀ ▀███▀ ▀███▀ ██▄▄▄ ██▄▄▄ ██ ██   AGPL-3.0-or-later - see LICENSE

package handlers

import (
	"errors"
	"net/http"

	"github.com/gorilla/mux"

	"juggler/cmd/juggler/core"
)

// The workspace table over HTTP. A workspace is where a conversation's tools
// run; the table is the set of them this project has, and it is session state
// like the pinboard — shared by every viewer, not a per-device preference.
//
// These hang off SessionAPI rather than an API of their own because they need
// exactly what it already holds: the session manager and the broadcaster. Every
// mutation broadcasts the whole table, so a second window sees a workspace
// appear, become usable, and be finished with, without asking.

// workspaceID reads the id out of the route, refusing a malformed one rather
// than defaulting. An id arrives from a client that was given it by this server,
// so one that does not look like an id is a bug at the other end.
func workspaceID(w http.ResponseWriter, r *http.Request) (string, bool) {
	id := mux.Vars(r)["workspaceId"]
	if id == "" {
		WriteError(w, r, http.StatusBadRequest, "workspace id is required")
		return "", false
	}
	return id, true
}

// workspaceStatus maps a manager error to the status it deserves: an unknown id
// is a request about something that is not there, anything else is a refusal of
// a well-formed request (a closed workspace, a table that is full).
func workspaceStatus(err error) int {
	if errors.Is(err, core.ErrWorkspaceNotFound) {
		return http.StatusNotFound
	}
	return http.StatusBadRequest
}

// broadcastWorkspaces publishes the whole table after an edit.
//
// The whole table rather than a diff, on the pinboard's reasoning: it is a
// short, wholly-owned list, and shipping it entire is what lets every viewer
// converge without replaying anyone's operations. It is also what carries the
// provisioning → ready flip to a window that is only watching.
func (api *SessionAPI) broadcastWorkspaces() {
	if api.broadcaster == nil {
		return
	}
	api.broadcaster.BroadcastWorkspacesChanged(api.manager().ListWorkspaces())
}

// HandleListWorkspaces returns every registered workspace.
//
// The default workspace — the project itself — is deliberately not among them.
// It has no row, every client already knows the project path, and putting it in
// the list would make "is this workspace registered" and "is this the project"
// the same question.
// Listing is also when the table's availability is re-checked, and the only
// time it is: a place can go while the app runs, and nothing else looks. The
// broadcast is conditional on something having actually moved, so an unchanged
// list stays a read.
func (api *SessionAPI) HandleListWorkspaces(w http.ResponseWriter, r *http.Request) {
	if api.manager().RefreshWorkspaceAvailability() {
		api.broadcastWorkspaces()
	}
	WriteJSON(w, r, 0, map[string]any{"workspaces": api.manager().ListWorkspaces()})
}

// HandleRegisterWorkspace puts a workspace on the table, in the provisioning
// state unless the caller says otherwise, and answers with the row as stored —
// including the id the server assigned it.
//
// It is registered before it is built. A provision that dies half way through
// then leaves a row describing what was started, instead of an unrecorded
// half-made worktree for the user to find months later.
func (api *SessionAPI) HandleRegisterWorkspace(w http.ResponseWriter, r *http.Request) {
	req, ok := DecodeJSON[core.Workspace](w, r)
	if !ok {
		return
	}
	ws, err := api.manager().RegisterWorkspace(req)
	if err != nil {
		WriteError(w, r, http.StatusBadRequest, err.Error())
		return
	}
	WriteJSON(w, r, http.StatusOK, map[string]any{"workspace": ws})
	api.broadcastWorkspaces()
}

// HandleUpdateWorkspace applies a patch to one workspace: its label, its root,
// its state, or a key or two of the provider's own record of what it has built.
//
// A patch rather than a whole-row PUT, and meta merged key by key, because the
// heaviest user of this endpoint is a provider checkpointing its progress
// mid-provision. It must be able to write "the tree exists now" without
// restating a row it has not finished building, and without overwriting what a
// second window wrote a moment earlier.
func (api *SessionAPI) HandleUpdateWorkspace(w http.ResponseWriter, r *http.Request) {
	id, ok := workspaceID(w, r)
	if !ok {
		return
	}
	patch, ok := DecodeJSON[core.WorkspacePatch](w, r)
	if !ok {
		return
	}
	ws, err := api.manager().UpdateWorkspace(id, patch)
	if err != nil {
		WriteError(w, r, workspaceStatus(err), err.Error())
		return
	}
	WriteJSON(w, r, http.StatusOK, map[string]any{"workspace": ws})
	api.broadcastWorkspaces()
}

// HandleCloseWorkspace tombstones a workspace. The row stays and the id keeps
// resolving, so a conversation still bound to it is told it was closed rather
// than told it is unknown — which is the message reserved for a binding that is
// genuinely stale.
func (api *SessionAPI) HandleCloseWorkspace(w http.ResponseWriter, r *http.Request) {
	id, ok := workspaceID(w, r)
	if !ok {
		return
	}
	ws, err := api.manager().CloseWorkspace(id)
	if err != nil {
		WriteError(w, r, workspaceStatus(err), err.Error())
		return
	}
	WriteJSON(w, r, http.StatusOK, map[string]any{"workspace": ws})
	api.broadcastWorkspaces()
}

// HandleUnregisterWorkspace removes a row outright — what rolling back a
// provision does, and the one case where forgetting is right: the workspace was
// never built, so nobody can have been bound to it.
func (api *SessionAPI) HandleUnregisterWorkspace(w http.ResponseWriter, r *http.Request) {
	id, ok := workspaceID(w, r)
	if !ok {
		return
	}
	if err := api.manager().UnregisterWorkspace(id); err != nil {
		WriteError(w, r, http.StatusInternalServerError, err.Error())
		return
	}
	WriteJSON(w, r, http.StatusOK, map[string]any{"ok": true})
	api.broadcastWorkspaces()
}

// HandleReorderWorkspaces rewrites the order the table is held in.
//
// The table's order is what decides between two boxes drawn in the same place —
// two of them with no conversation between them to sit behind — so it is part
// of the arrangement a user drags the sidebar into, and has to be recorded like
// the rest of it. Ids the table does not have are ignored, and rows the caller
// did not name keep their order behind the ones it did.
func (api *SessionAPI) HandleReorderWorkspaces(w http.ResponseWriter, r *http.Request) {
	req, ok := DecodeJSON[struct {
		IDs []string `json:"ids"`
	}](w, r)
	if !ok {
		return
	}
	list, err := api.manager().ReorderWorkspaces(req.IDs)
	if err != nil {
		WriteError(w, r, workspaceStatus(err), err.Error())
		return
	}
	WriteJSON(w, r, http.StatusOK, map[string]any{"workspaces": list})
	api.broadcastWorkspaces()
}

// HandleClaimWorkspaceReconcile answers whether this viewer is the one to
// reconcile the table against what is actually on disk, and answers yes at most
// once per run of the server.
//
// The server cannot do it: reconciling means asking each provider what it can
// find, and providers are extensions, which live in the browser. But there is no
// leader among clients, so without this every open window would run it — and two
// windows would race each other's destructive git commands over the same trees.
// It is a POST rather than a GET because the claim is spent by asking.
func (api *SessionAPI) HandleClaimWorkspaceReconcile(w http.ResponseWriter, r *http.Request) {
	WriteJSON(w, r, http.StatusOK, map[string]any{"reconcile": api.manager().ClaimWorkspaceReconcile()})
}
